### PR TITLE
Add public `managesAudioSession` flag and fix `willSpeak` logic

### PR DESCRIPTION
**Pull request by azarovalex**
This PR fixes two issues in `MapboxSpeechSynthesizer`:
1. `managesAudioSession` is a new flag in `MapboxSpeechSynthesizer ` that defines who controls the shared AVAudioSession – SDK or Application. Default value is `true` that means that SDK manages the session, as it used to be before. If application has background music they want to manage the audio session themselves, so they should set this flag to `false`.
Without setting this flag to `false` we will be stopping the `AVAudioSession` after every voice instruction and this will stop application music as well.

2. We used to call `willSpeak` before downloading and caching the instruction, not before playing it. `MapboxSpeechSynthesizer` pre-caches first three voice instruction on initialization and this leads to bug https://github.com/mapbox/mapbox-navigation-ios/issues/3562. I fixed it by moving `willSpeak` call right before playing the instruction.

TODO:
- [x] update changelog

### CODE REVIEW

File: Sources/MapboxNavigation/MultiplexedSpeechSynthesizer.swift [hunk 1]
```diff
@@ -40,6 +40,11 @@ open class MultiplexedSpeechSynthesizer: SpeechSynthesizing {
         speechSynthesizers.forEach { $0.locale = locale }
     }
     
+    /// Controls if this speech synthesizer is allowed to manage the shared `AVAudioSession`.
+    /// Set this field to `false` if you want to manage the session yourself, for example if your app has background music.
+    /// Default value is`true`.
+    public var managesAudioSession = true
```
Comment (Udumft) [inline Sources/MapboxNavigation/MultiplexedSpeechSynthesizer.swift:46]: This has to be a computed property which will forward `newValue` to underlying speech synthesizers.

Comment (azarovalex) [reply Sources/MapboxNavigation/MultiplexedSpeechSynthesizer.swift:46]: Good idea, thanks

**Comment by S2Ler** [comment]
> # Breaking Changes in MapboxNavigation
> # Breaking API Changes
> ## `InstructionsCardCell`
> * _removed_ method: `layoutSubviews()` in `InstructionsCardCell`
> * _removed_ var: `container` in `InstructionsCardCell`

I'm fixing this.

File: Sources/MapboxNavigation/MapboxSpeechSynthesizer.swift [hunk 2]
```diff
@@ -35,6 +35,11 @@ open class MapboxSpeechSynthesizer: NSObject, SpeechSynthesizing {
      */
     public var audioPlayer: AVAudioPlayer?
     
+    /// Controls if this speech synthesizer is allowed to manage the shared `AVAudioSession`.
+    /// Set this field to `false` if you want to manage the session yourself, for example if your app has background music.
+    /// Default value is`true`.
```
Comment (bamx23) [inline Sources/MapboxNavigation/MapboxSpeechSynthesizer.swift:40]:
```suggestion
    /// Default value is `true`.
```

File: Sources/MapboxNavigation/MultiplexedSpeechSynthesizer.swift [hunk 3]
```diff
@@ -40,6 +40,18 @@ open class MultiplexedSpeechSynthesizer: SpeechSynthesizing {
         speechSynthesizers.forEach { $0.locale = locale }
     }
     
+    /// Controls if this speech synthesizer is allowed to manage the shared `AVAudioSession`.
+    /// Set this field to `false` if you want to manage the session yourself, for example if your app has background music.
+    /// Default value is`true`.
+    public var managesAudioSession: Bool {
+        get {
+            speechSynthesizers.reduce(true, { $0 && $1.managesAudioSession })
```
Comment (bamx23) [inline Sources/MapboxNavigation/MultiplexedSpeechSynthesizer.swift:48]: NIT: What do you think about such approach?
```suggestion
            !speechSynthesizers.contains(where: { !$0.managesAudioSession })
```

File: Sources/MapboxNavigation/MultiplexedSpeechSynthesizer.swift [hunk 4]
```diff
@@ -40,6 +40,18 @@ open class MultiplexedSpeechSynthesizer: SpeechSynthesizing {
         speechSynthesizers.forEach { $0.locale = locale }
     }
     
+    /// Controls if this speech synthesizer is allowed to manage the shared `AVAudioSession`.
+    /// Set this field to `false` if you want to manage the session yourself, for example if your app has background music.
+    /// Default value is`true`.
```
Comment (bamx23) [inline Sources/MapboxNavigation/MultiplexedSpeechSynthesizer.swift:45]:
```suggestion
    /// Default value is `true`.
```

File: Sources/MapboxNavigation/SpeechSynthesizing.swift [hunk 5]
```diff
@@ -18,6 +18,10 @@ public protocol SpeechSynthesizing: AnyObject {
     var isSpeaking: Bool { get }
     /// Locale setting to vocalization. This locale will be used as 'default' if no specific locale is passed for vocalizing each individual instruction.
     var locale: Locale? { get set }
+    /// Controls if this speech synthesizer is allowed to manage the shared `AVAudioSession`.
+    /// Set this field to `false` if you want to manage the session yourself, for example if your app has background music.
+    /// Default value is`true`.
```
Comment (bamx23) [inline Sources/MapboxNavigation/SpeechSynthesizing.swift:23]:
```suggestion
    /// Default value is `true`.
```

File: Sources/MapboxNavigation/SystemSpeechSynthesizer.swift [hunk 6]
```diff
@@ -30,6 +30,11 @@ open class SystemSpeechSynthesizer: NSObject, SpeechSynthesizing {
     
     public var locale: Locale? = Locale.autoupdatingCurrent
     
+    /// Controls if this speech synthesizer is allowed to manage the shared `AVAudioSession`.
+    /// Set this field to `false` if you want to manage the session yourself, for example if your app has background music.
+    /// Default value is`true`.
```
Comment (bamx23) [inline Sources/MapboxNavigation/SystemSpeechSynthesizer.swift:35]:
```suggestion
    /// Default value is `true`.
```

File: Tests/MapboxNavigationTests/NavigationViewControllerTestDoubles.swift [hunk 7]
```diff
@@ -17,6 +17,7 @@ class SpeechSynthesizerStub: SpeechSynthesizing {
     var volume: Float = 1.0
     var isSpeaking: Bool = false
     var locale: Locale? = Locale.autoupdatingCurrent
+    var managesAudioSession: Bool = false
```
Comment (bamx23) [inline Tests/MapboxNavigationTests/NavigationViewControllerTestDoubles.swift:20]: Why is it false in tests?

**Review by bamx23 — APPROVED** [review]

File: Sources/MapboxNavigation/MultiplexedSpeechSynthesizer.swift (hunk 3 above)
Comment (azarovalex) [reply Sources/MapboxNavigation/MultiplexedSpeechSynthesizer.swift:48]: Agree, might be easier to understand

**Pull request merged**